compiler/ras/ILValidator.cpp

```diff
@@ -0,0 +1,340 @@
+      FAIL();
+      }
+   }
+
+void TR::ILValidator::validityRule(Location &location, bool condition, const char *formatStr, ...)
+   {
+   if (!condition)
+      {
+      _isValidSoFar = false;
+      TR::Node *node = location.currentNode();
+      printDiagnostic("*** VALIDATION ERROR ***\nNode: %s n%dn\nMethod: %s\n", node->getOpCode().getName(), node->getGlobalIndex(), comp()->signature());
+      va_list args;
+      va_start(args, formatStr);
+      vprintDiagnostic(formatStr, args);
+      va_end(args);
+      printDiagnostic("\n");
+      FAIL();
+      }
+   }
+
+void TR::ILValidator::updateNodeState(Location &newLocation)
+   {
+   TR::Node  *node = newLocation.currentNode();
+   NodeState &state = _nodeStates[node];
+   if (node->getReferenceCount() == state._futureReferenceCount)
+      {
+      // First occurrence -- do some bookkeeping
+      //
+      if (node->getReferenceCount() == 0)
+         {
+         validityRule(newLocation, node->getOpCode().isTreeTop(), "Only nodes with isTreeTop opcodes can have refcount == 0");
+         }
+      else
+         {
+         _liveNodes.add(node);
+         }
+      }
+
+   if (_liveNodes.contains(node))
+      {
+      validityRule(newLocation, state._futureReferenceCount >= 1, "Node already has reference count 0");
+      if (--state._futureReferenceCount == 0)
+         {
+         _liveNodes.remove(node);
+         }
+      }
+   else
+      {
+      validityRule(newLocation, node->getOpCode().isTreeTop(), "Node has already gone dead");
+      }
+
+   if (isLoggingEnabled())
+      {
+      static const char *traceLiveNodesDuringValidation = feGetEnv("TR_traceLiveNodesDuringValidation");
+      if (traceLiveNodesDuringValidation && !_liveNodes.isEmpty())
+         {
+         traceMsg(comp(), "    -- Live nodes: {");
+         char *separator = "";
+         for (LiveNodeWindow::Iterator lnwi(_liveNodes); lnwi.currentNode(); ++lnwi)
+            {
+            traceMsg(comp(), "%sn%dn", separator, lnwi.currentNode()->getGlobalIndex());
+            separator = ", ";
+            }
+         traceMsg(comp(), "}\n");
+         }
+      }
+
+   }
+
+static bool hasGlRegDeps(TR::Node* node)
+   {
+   for (auto i = 0; i < node->getNumChildren(); ++i)
+      {
+      if (node->getChild(i)->getOpCodeValue() == TR::GlRegDeps)
+         {
+         return true;
+         }
+      }
+   return false;
+   }
+
+bool TR::ILValidator::treesAreValid(TR::TreeTop *start, TR::TreeTop *stop)
+   {
+   checkSoundness(start, stop);
+
+   for (PostorderNodeOccurrenceIterator iter(start, _comp, "VALIDATOR"); iter != stop; ++iter)
+      {
+      updateNodeState(iter);
+
+      // General node validation
+      //
+      validateNode(iter);
+
+      //
+      // Additional specific kinds of validation
+      //
+
+      TR::Node *node = iter.currentNode();
+      if (node->getOpCodeValue() == TR::BBEnd)
+         {
+         // Determine whether this is the end of an extended block
+         //
```

Comment:
There would be an opportunity here to catch another kind of oops - where you have broken a block and failed to extend when it makes sense to do so. This would be more of a warning, but it could be a nice extra in another iteration to check if there is a single successor, single predecessor relationship and block extension has been run then we should expect the next block to be marked an extension.